### PR TITLE
Fix "Plot surface/contour from group"

### LINKED ISSUE #17896: PlotSurface and contour from group broken
### Actual behavior

After the dialog box you get this error
image
### Steps to reproduce the behavior

Right click on a group with more than 2 members.
Click ok on the dialog box
### Platforms affected

All

### PR DESCRIPTION
**Pull request by tom-perkins**
Refactor the code for plotting surface/contour plots from a WorkspaceGroup to use `HistogramData`.
This ensures that it now works and does not throw an exception for histogram workspaces.

To avoid making a 3D plot of a ragged workspace, the input workspaces are required to all have the same X data length.

**To test:**

Use the following script:

``` python
def create_sample_group(n_items, histo, name):
    to_group = list()
    xData = range(200, 211, 1) if histo else range(200, 210, 1)
    for i in range(n_items):
        ws_name = name + str(i)
        to_group.append(ws_name)
        CreateWorkspace(DataX=xData, DataY=[i+1]*20, NSpec=2, OutputWorkspace=ws_name)
        AddSampleLog(Workspace=ws_name, LogName="MyLog", LogText=str((i+1)*5), LogType='Number')
    GroupWorkspaces(InputWorkspaces=','.join(to_group), OutputWorkspace="gp_"+name)

create_sample_group(n_items=10, histo=True, name="histo")
create_sample_group(n_items=10, histo=False, name="points")
```

Right-click on each group and choose "Plot surface from group" or "Plot contour from group". Check that you do not see the exception in the issue when making the plot.

If you try this with other workspace groups, note that all workspaces in the group must have the same X data length (or an error message will be displayed).

No tests as this is in MantidPlot.

Fixes #17896.

[Release notes](https://github.com/mantidproject/mantid/blob/64e93cf6d9a04dc34bc8e2ae9bb2a280e26695e5/docs/source/release/v3.9.0/ui.rst#bugs-resolved) have been updated.

---
#### Reviewer

Please comment on the following ([full description](http://www.mantidproject.org/Individual_Ticket_Testing)):
##### Code Review
- [x] Is the code of an acceptable quality?
- [x] Does the code conform to the [coding standards](http://www.mantidproject.org/Coding_Standards)? Is it well structured with small focussed classes/methods/functions?
- [ ] Are there unit/system tests in place? Are the unit tests small and test the a class in isolation?
- [x] If there are changes in the release notes then do they describe the changes appropriately?
##### Functional Tests
- [x] Do changes function as described? Add comments below that describe the tests performed?
- [x] How do the changes handle unexpected situations, e.g. bad input?
- [ ] Has the relevant documentation been added/updated?
- [ ] Is user-facing documentation written in a user-friendly manner?
- [x] Has developer documentation been updated if required?
- Does everything look good? Comment with the ship it emoji but don't merge. A member of `@mantidproject/gatekeepers` will take care of it.

### CODE REVIEW

File: MantidPlot/src/Mantid/MantidGroupPlotGenerator.cpp [hunk 1]
```diff
@@ -313,3 +272,104 @@ std::string MantidGroupPlotGenerator::validatePlotOptions(
   }
   return err.str();
 }
+
+/**
+ * Generates X axis title for graph based on first workspace in group
+ */
```
Comment (gvardany) [inline MantidPlot/src/Mantid/MantidGroupPlotGenerator.cpp]: `@param` and `@return` are missing

File: MantidPlot/src/Mantid/MantidGroupPlotGenerator.cpp [hunk 2]
```diff
@@ -313,3 +272,104 @@ std::string MantidGroupPlotGenerator::validatePlotOptions(
...
+
+/**
+ * Test if all workspaces in the group have the same X data for the given
+ * spectrum.
+ * (At the moment just tests size of X data)
+ * Precondition: wsGroup contains only MatrixWorkspaces
+ *
+ * @param wsGroup :: [input] Group to test
+ * @param index :: [input] Index of spectrum to test
+ * @return :: True if X data same, else false.
```
Comment (gvardany) [inline MantidPlot/src/Mantid/MantidGroupPlotGenerator.cpp]: `@throw` is missing

**Review by gvardany — CHANGES_REQUESTED** [review]
Works as expected. Could you please complete the doxygen in 2 cases(see inline) ?

**Comment by tom-perkins** [comment]
Thanks for the review, I've corrected the doxygen comments in these two places.

**Review by gvardany — APPROVED** [review]

**Comment by gvardany** [comment]
Thanks, happy to :shipit:

**Pull request merged**